common/src/jni/main/cpp/conscrypt/native_crypto.cc

```diff
@@ -3658,6 +3658,14 @@ static jint evp_aead_ctx_op_buf(JNIEnv* env, jlong evpAeadRef, jbyteArray keyArr
     inBuf += in_position;
     outBuf += out_position;
 
+    uint8_t* outBufEnd = outBuf + out_limit - out_position;
+    uint8_t* inBufEnd = inBuf + in_limit - in_position;
+    if (outBufEnd >= inBuf && inBufEnd >= outBuf) { // We have an overlap
+        std::vector<uint8_t> inVector(inBuf, inBuf + in_limit - in_position);
+        return evp_aead_ctx_op_common(env, evpAeadRef, keyArray, tagLen, outBuf, nonceArray, inVector.data(), aadArray,
```

Comment:
inBuf is just a pointer, so you can assign inVector.data() to it here and and avoid the duplicate return statements.

I'm slightly concerned that we're allocating the copy on the stack here too.  The code we're replacing  allocates its copies in the Java heap so can handle much larger buffers.

Reply:
Yeah, I think we should definitely allocate the buffer copy, e.g.
```
size_t inSize = in_limit - in_position;
std::unique_ptr<uint8_t[]> inCopy(new uint8_t[inSize]);
if (inCopy.get() == nullptr) {
    conscrypt::jniutil::throwOutOfMemory(env, "Unable to allocate buffer");
    return 0;
}
std::memcpy(inCopy.get(), inBuf, inSize);
inBuf = inCopy.get();
```

Reply:
the multiple return statements were a way to try to avoid having to create this inCopy in every single execution and only use it in the overlapping case. Would it be ok to always create a copy of inBuf?